twister: passing the build dir to device serial pty

enjiamai opened the pull request:
Passing the zephyr build directory to the script run by
--device-serial-pty, it makes function of  the device serial
pty more flexible.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

andyross reviewed (dismissed):
This makes sense.  Is there a place to document this behavior?  (Maybe not, I know I was confused by the serial-pty feature[1] when I first looked at it).

[1] Which involves no serial ports, nor pseudoterminal devices!  It's a feature for things that are NOT serial ports or ptys, I guess?  pseusopseudoptys?

andyross commented:
FWIW: this feature obsoletes a bunch of the logic in cavstwist.sh that is designed to pass the signed rimage file reliably between the two script phases.  If you want, feel free to simplify that too.

enjiamai commented:
> This makes sense. Is there a place to document this behavior? (Maybe not, I know I was confused by the serial-pty feature[1] when I first looked at it).
> 
> [1] Which involves no serial ports, nor pseudoterminal devices! It's a feature for things that are NOT serial ports or ptys, I guess? pseusopseudoptys?

Hi @andyross , I also get confused for a period of time. Anyway, spawn you great script here works better, I think. 
We don't worry about how to get the output from west. The only constraint is we need the "build directory path" for our script.

enjiamai commented:
> FWIW: this feature obsoletes a bunch of the logic in cavstwist.sh that is designed to pass the signed rimage file reliably between the two script phases. If you want, feel free to simplify that too.

Ok, I will give it a try :).

chen-png commented on scripts/pylib/twister/twisterlib.py:
```diff
@@ -730,8 +730,12 @@ def handle(self):
-                ser_pty_process = subprocess.Popen(re.split(',| ', serial_pty), stdout=master, stdin=master, stderr=master)
+                ser_pty_process = subprocess.Popen(serial_cmd, stdout=master, stdin=master, stderr=master)
             except subprocess.CalledProcessError as error:
                 logger.error("Failed to run subprocess {}, error {}".format(serial_pty, error.output))
```
serial_pty --> serial_cmd

enjiamai replied:
Hi @chen-png,  it's serial_cmd already.  I guess you mean the removed old one. :)

chen-png replied:
I mean you should use serial_cmd too in the looger.error.

enjiamai replied:
ok, got your point, thank for correcting it.

enjiamai replied:
@chen-png , got it fixed, thanks!

enjiamai committed:
twister: passing the build dir to device serial pty

Passing the zephyr build directory to the script run by
--device-serial-pty, it makes the feature of the device serial
pty more flexible.

Signed-off-by: Enjia Mai <enjia.mai@intel.com>

nashif commented on scripts/pylib/twister/twisterlib.py:
```diff
@@ -730,10 +730,14 @@ def handle(self):
         if serial_pty:
             master, slave = pty.openpty()
+            serial_cmd = re.split(',| ', serial_pty)
+            # We try to pass the build directory to the device serial pty for specially purpose
```
> for specially purpose
you need to be more specific here.

enjiamai replied:
Thanks for the review!  @nashif , do you think the updated description is appropriate or not?

enjiamai commented:
hold down please, double check.

enjiamai commented:
> hold down please, double check.

Checked. No problem.

enjiamai commented:
Hi @nashif , thanks for your help for implementing the runner. This is not necessary now. But I think there is still some benefit to have this.  How do you think? Thanks!

hakehuang commented on scripts/pylib/twister/twisterlib.py:
```diff
@@ -749,10 +749,19 @@ def handle(self):
+            # one script. According to twister's working flow, the script is better to be put in
+            # --device-serial-pty script in these cases. So like what we pass to --west-flash script,
+            # we also pass also build directory to pty script, in order to get more flexibility.
+            serial_cmd.append(self.build_dir)
```
do we have document for this --device-serial-pty option? currently there are only one simple help in twister 
```
twister --device-testing
                        --device-serial-pty <script>
```
with this change the script shall take the build_dir as its first parameter, better to update this helper

enjiamai replied:
ok, @hakehuang, good suggestion. I could see where is the good place to add this.

hakehuang commented on scripts/pylib/twister/twisterlib.py:
```diff
@@ -749,10 +749,19 @@ def handle(self):
+            # intel_adsp_cavs series, their flash script and log capturing script might be put into
+            # one script. According to twister's working flow, the script is better to be put in
+            # --device-serial-pty script in these cases. So like what we pass to --west-flash script,
+            # we also pass also build directory to pty script, in order to get more flexibility.
```
suggestion` # the pty script will ahve build directory as its first paramenter by derault, in order to get more flexibility.`

hakehuang replied:
```suggestion
            #  the pty script will ahve build directory as its first paramenter by derault, in order to get more flexibility.
```

enjiamai commented on scripts/pylib/twister/twisterlib.py:
```diff
@@ -749,10 +749,19 @@ def handle(self):
+            # intel_adsp_cavs series, their flash script and log capturing script might be put into
+            # one script. According to twister's working flow, the script is better to be put in
+            # --device-serial-pty script in these cases. So like what we pass to --west-flash script,
+            # we also pass also build directory to pty script, in order to get more flexibility.
```
Hi @hakehuang, thanks for the suggestion! Beside what you mentioned,  I also updated the descriptions to fit the current status. could you take a look at it again, thanks.
